### PR TITLE
feature/issue-428 defer to poisson for bad parameters in the negative binomial

### PR DESCRIPTION
**Pull request by thelseraphim**
#### Submisison Checklist

- [ ] Run unit tests: `./runTests.py test/unit`
- [ ] Run cpplint: `make cpplint`
- [ ] Declare copyright holder and open-source license: see below

#### Summary:

In the mean/inverse-overdispersion parametrization of the negative binomial distribution we get overflows for ultra-large values of the inverse over-dispersion. This results in counter-intuitive or extremely wrong behavior (positive values for log probability) due to cancellation between extremely large values. We add a test for large values to catch that particular overflow and defer to poisson_log when the over-dispersion is near zero.

Ideally we would never use the inverse-overdispersion-tending-to-infinity area of the negative binomial parameters since, but since this value can be sampled or bizarrely initialized we can't quite avoid dealing with this.

#### Intended Effect:
Fix the bug here: https://github.com/stan-dev/math/issues/428
also relevant is this closed issue which duplicates it https://github.com/stan-dev/math/issues/463

#### How to Verify:
can run the test here `(ProbDistributionsNegBinomial, extreme_values)`: `./runTests.py ./test/unit/math/prim/scal/prob/neg_binomial_2_test.cpp`

decide whether this is a reasonable fallback.

#### Side Effects:
Changes values which were almost certainly nonsense before.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Columbia University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
- Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

### CODE REVIEW

File: stan/math/prim/scal/prob/neg_binomial_2_log.hpp [hunk 1]
```diff
@@ -95,6 +96,7 @@ namespace stan {
       for (size_t i = 0; i < len_np; ++i)
         n_plus_phi[i] = n_vec[i] + phi__[i];
 
+      using std::cout;
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/neg_binomial_2_log.hpp]: We can't use `cout` or `cin` or `cerr`---it makes it impossible to wrap with R, Python or any other process that controls its own I/O.

File: stan/math/prim/scal/prob/neg_binomial_2_log.hpp [hunk 2]
```diff
@@ -107,6 +109,11 @@ namespace stan {
         if (include_summand<propto, T_precision>::value)
           logp += lgamma(n_plus_phi[i]);
 
+        // if phi is large we probably overflow, defer to Poisson:
+        if (phi__[i] > 1e12) {
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/neg_binomial_2_log.hpp]: Is this because as phi gets big, the negative binomial is Poisson like?  If that's right, can you mention that it approaches a Poisson and replace "defer" with "delegate".

File: stan/math/prim/scal/prob/neg_binomial_2_log.hpp [hunk 3]
```diff
@@ -107,6 +109,11 @@ namespace stan {
...
+          logp = poisson_log(n_vec[i], mu__[i]);
+        }
+
         if (!is_constant_struct<T_location>::value)
```
Comment (bob-carpenter) [inline stan/math/prim/scal/prob/neg_binomial_2_log.hpp]: Will the derivatives w.r.t. phi still work?

File: test/unit/math/prim/scal/prob/neg_binomial_2_test.cpp [hunk 4]
```diff
@@ -202,3 +202,14 @@ TEST(ProbDistributionsNegBinomial, chiSquareGoodnessFitTest4) {
 
   EXPECT_TRUE(chi < quantile(complement(mydist, 1e-6)));
 }
+
```
Comment (bob-carpenter) [inline test/unit/math/prim/scal/prob/neg_binomial_2_test.cpp]: There should be a derivative test, too, in rev.  I'm not sure how to plug this into the testing framework, but @syclik will.

File: stan/math/prim/scal/prob/neg_binomial_2_log.hpp (hunk 3 above)
Comment (thelseraphim) [inline stan/math/prim/scal/prob/neg_binomial_2_log.hpp]: They should be ok, digamma grows like log so is quite well-behaved for large inputs, but I agree that it ought to have a test.

File: test/unit/math/prim/scal/prob/neg_binomial_2_test.cpp (hunk 4 above)
Comment (thelseraphim) [inline test/unit/math/prim/scal/prob/neg_binomial_2_test.cpp]: tiny update on this, I have a reasonable test for the derivatives which reveals an error in the derivative with respect to *mu* when phi is large. I'll flag the PR again when it's fixed

**Comment by thelseraphim** [comment]
I cleaned this up a bit and added a test for the derivatives in rev, covering extreme values of phi (check that they're not nan and aren't too terribly far from the finite differences). Hopefully ready to go.

**Comment by bob-carpenter** [comment]
Thanks much for adding derivative tests.  This is good to merge when the tests finish.
